ratpack-core/src/test/groovy/ratpack/ssl/HttpsTruststoreSpec.groovy

```diff
@@ -0,0 +1,159 @@
+      sslClientAuth(true)
+    }
+
+    handlers {
+      get("foo") {
+        render "SSL VERIFIED"
+      }
+    }
+
+    when:
+    requestSpec { RequestSpec rs ->
+      rs.sslContext {
+        return SSLContexts.sslContext(
+          HttpsTruststoreSpec.getResource("client_dummy.keystore"), "password",
+          HttpsTruststoreSpec.getResource("client_dummy.truststore"), "password")
+      }
+    }
+
+    then:
+    def address = applicationUnderTest.address
+    address.scheme == "https"
+    getText("foo") == "SSL VERIFIED"
+  }
+
+  def "throw exception when client provides no or wrong certificate and client ssl authentication is required"() {
+    given:
+    serverConfig {
+      ssl SSLContexts.sslContext(
+        HttpsTruststoreSpec.getResource("server_dummy.keystore"), "password",
+        HttpsTruststoreSpec.getResource("server_dummy.truststore"), "password")
+      sslClientAuth true  // require client ssl authentication
+    }
+
+    handlers {
+      get("foo") {
+        render "SSL VERIFIED"
+      }
+    }
+
+    when: "client does not provide valid certificate"
+    requestSpec { RequestSpec rs ->
+      rs.sslContext {
+        return SSLContexts.sslContext(
+          HttpsTruststoreSpec.getResource("dummy.keystore"), "password",
+          HttpsTruststoreSpec.getResource("client_dummy.truststore"), "password")
+      }
+    }
+    get("foo")
+
+    then:
+    UncheckedIOException ex = thrown()
+    ex.getCause() instanceof SSLProtocolException || ex.getCause() instanceof ClosedChannelException
+
+    when: "client does not provide truststore"
+    resetRequest()
+    requestSpec { RequestSpec rs ->
+      rs.sslContext {
+        return SSLContexts.sslContext(
+          HttpsTruststoreSpec.getResource("dummy.keystore"), "password")
+      }
+    }
+    get("foo")
+
+    then:
+    ex = thrown()
+    ex.getCause() instanceof SSLHandshakeException || ex.getCause() instanceof ClosedChannelException
+  }
+
+  def "throw exception when server does not provide valid keystore"() {
+    given:
+    serverConfig {
+      ssl SSLContexts.sslContext(
+        HttpsTruststoreSpec.getResource("dummy.keystore"), "password",
+        HttpsTruststoreSpec.getResource("server_dummy.truststore"), "password")
+      sslClientAuth true  // require client ssl authentication
+    }
+
+    handlers {
+      get("foo") {
+        render "SSL VERIFIED"
+      }
+    }
+
+    when: "client provides valid certificate"
+    requestSpec { RequestSpec rs ->
+      rs.sslContext {
+        return SSLContexts.sslContext(
+          HttpsTruststoreSpec.getResource("client_dummy.keystore"), "password",
+          HttpsTruststoreSpec.getResource("client_dummy.truststore"), "password")
+      }
+    }
+    get("foo")
+
+    then:
+    UncheckedIOException ex = thrown()
+    ex.getCause() instanceof SSLHandshakeException || ex.getCause() instanceof ClosedChannelException
+  }
+
+  def "throw exception when server does not provide truststore"() {
```

Comment:
Exception is thrown during the handshake. Before client -> server call we do not know if keystore or trusture contains valid and required certificates or key pairs.